### PR TITLE
KFSPTS-10604: Allow expired accounts on POA documents

### PR DESCRIPTION
**Pull request by cah292**
This is a fixed version of the code from the #495 and #501 PRs.

NOTE: There are PRs in both cu-kfs and nonprod-sql for this feature. Please make sure both are good to go before merging them.

I also went ahead and applied the SQL to UTEST, in case the unit tests need the changes present in order to boot and run.

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderForm.java [hunk 1]
```diff
@@ -194,6 +198,61 @@ protected boolean canVoid() {
...
+        int itemCount = 0;
+        for (PurApItem item : ((PurchasingAccountsPayableDocument) getDocument()).getItems()) {
+            populateAccountingLine(item.getNewSourceLine(), KFSPropertyConstants.DOCUMENT + "." + KFSPropertyConstants.ITEM + "[" + itemCount + "]."
+                    + KFSPropertyConstants.NEW_SOURCE_LINE, parameterMap);
+
+            int sourceLineCount = 0;
+            for (PurApAccountingLine purApLine : item.getSourceAccountingLines()) {
+                populateAccountingLine(purApLine, KFSPropertyConstants.DOCUMENT + "." + KFSPropertyConstants.ITEM + "[" + itemCount + "]."
+                        + KFSPropertyConstants.SOURCE_ACCOUNTING_LINE + "[" + sourceLineCount + "]", parameterMap);
+                sourceLineCount += 1;
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderForm.java]: to be consistent, should we use sourceLineCount++

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderAction.java [hunk 2]
```diff
@@ -162,4 +174,57 @@ public ActionForward cancel(ActionMapping mapping, ActionForm form,
...
+             */
+            processAccountingLineOverrides(null, getItemAccountingAddLines(poaDocument));
+        } else {
+            super.processAccountingLineOverrides(transForm);
+        }
+    }
+
+    protected List<PurApAccountingLine> getItemSourceAccountingLines(PurchaseOrderAmendmentDocument document) {
+        return getAccountingLinesFromItems(document,
+                (itemStream) -> itemStream.flatMap(this::getItemSourceAccountingLinesAsStream));
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderAction.java]: can this lambda expression and the similar ones below be refactored to help with understanding, maybe some different naming.

File: src/main/java/edu/cornell/kfs/module/purap/document/service/impl/CuPurchaseOrderServiceImpl.java [hunk 3]
```diff
@@ -190,4 +189,24 @@ private PurchaseOrderDocument copyNotesAndAttachmentsToPO(RequisitionDocument re
...
+     * 
+     * @see org.kuali.kfs.module.purap.document.service.impl.PurchaseOrderServiceImpl#createPurchaseOrderDocumentFromSourceDocument(
+     * org.kuali.kfs.module.purap.document.PurchaseOrderDocument, java.lang.String)
+     */
+    @Override
+    protected PurchaseOrderDocument createPurchaseOrderDocumentFromSourceDocument(
+            PurchaseOrderDocument sourceDocument, String docType) throws WorkflowException {
+        PurchaseOrderDocument newDocument = super.createPurchaseOrderDocumentFromSourceDocument(sourceDocument, docType);
+        
+        List<PurApItem> items = (List<PurApItem>) newDocument.getItems();
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/document/service/impl/CuPurchaseOrderServiceImpl.java]: Good point, I will refactor accordingly to allow for simplifying or removing the in-code comments.

File: src/main/java/edu/cornell/kfs/module/purap/document/service/impl/CuPurchaseOrderServiceImpl.java [hunk 4]
```diff
@@ -190,4 +189,24 @@ private PurchaseOrderDocument copyNotesAndAttachmentsToPO(RequisitionDocument re
...
+        items.stream()
+                .flatMap((item) -> item.getSourceAccountingLines().stream())
+                .forEach((accountingLine) -> accountingLine.setOverrideCode(AccountingLineOverride.CODE.NONE));
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/document/service/impl/CuPurchaseOrderServiceImpl.java]: Accounting lines typically have an override code of "NONE" by default. I could filter by accounting lines with non-"NONE" codes to limit the scope of lines to be updated.

PURAP items only have source lines and no target lines, and PO documents do not appear to have target lines either.

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        List<PurApItem> items = poaDocument.getItemsActiveOnly();
+
+        PurchaseOrderDocument po = SpringContext.getBean(PurchaseOrderService.class).getCurrentPurchaseOrder(poaDocument.getPurapDocumentIdentifier());
+        List<PurApItem> poItems = po.getItems();
+
+        for (PurApItem item : items) {
+            String identifierString = item.getItemIdentifierString();
+
+            if (item.getItemTypeCode().equals(PurapConstants.ItemTypeCodes.ITEM_TYPE_ITEM_CODE)
+                    && item.getSourceAccountingLines() != null && item.getSourceAccountingLines().size() > 0) {
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java]: This is part of a larger code section that was copied from base KFS code. I would prefer to keep it as close to base code as possible, with the exception of the changes needed to implement the new feature.

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                    }
+                }
+            }
+        }
+
+        return valid;
+    }
+
+    // Copied and tweaked this superclass method, and increased its visibility.
+    protected boolean isItemChanged(PurApItem poaItem, List<PurApItem> poItems) {
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java]: The original method in base KFS was declared as private, so the "Override" annotation cannot be placed here.

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderAction.java (hunk 2 above)
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/document/web/struts/CuPurchaseOrderAction.java]: Ok, I will refactor this accordingly to improve readability.

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java]: ah, OK

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/impl/CuPurchaseOrderAmendmentAccountValidation.java]: ok

**Comment by cah292** [comment]
I've made some changes based on the initial round of feedback, to improve readability and to reduce the use of comments slightly. This is now ready for review again.

**Comment by cah292** [comment]
NOTE: For the unit test run from when I first opened this PR, it somehow took around an hour for Jenkins to finish running the tests. Not sure if it was an issue with Jenkins or with this code.

Before merging, please verify that the unit test run for my most recent changes actually completed within the expected amount of time.

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
This PR looks fine now, but the unit tests failed with "kfs-war didn't run" ... trying the unit tests again.

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
retest this please

**Comment by cah292** [comment]
retest this please

**Pull request merged**